File: src/net/cicada/event/api/EventCaller.java
package net.cicada.event.api;

import java.util.ArrayList;
import java.util.List;

public class EventCaller {
    private static final List<EventListener> listeners = new ArrayList<>();

    public static void register(EventListener... listeners) {
        EventCaller.listeners.addAll(List.of(listeners));
    }

    public static void register(List<EventListener> listeners) {
        EventCaller.listeners.addAll(listeners);
    }

    public static void call(Event e) {
        for (EventListener listener : listeners) {
            if (!listener.listen()) continue;
            listener.listen(e);
        }
    }
}

File: src/main/kotlin/cicada/client/event/EventCaller.kt
package cicada.client.event

@Suppress("UNUSED_EXPRESSION")
object EventCaller {
    private val listeners = mutableSetOf<EventListener>()

    internal operator fun plusAssign(listener: EventListener) {
        if (listener !in listeners)
            listeners += listener
    }

    internal fun call(event: Event) {
        if (event is CancelableEvent)
            event.reset()

        for (listener in listeners) {
            if (!listener.shouldListenEvents())
                continue

            listener.onEvent(event)
        }
    }
}